src/hotspot/share/gc/shenandoah/shenandoahFreeSet.cpp

```diff
@@ -437,29 +514,107 @@ HeapWord* ShenandoahFreeSet::try_allocate_in(ShenandoahHeapRegion* r, Shenandoah
     // available free segments within regions for more efficient searches for "good fit".
 
     // Record the remainder as allocation waste
+    size_t idx = r->index();
     if (req.is_mutator_alloc()) {
       size_t waste = r->free();
       if (waste > 0) {
         increase_used(waste);
         generation->increase_allocated(waste);
         _heap->notify_mutator_alloc_words(waste >> LogHeapWordSize, true);
       }
-    }
-
-    size_t num = r->index();
-    _collector_free_bitmap.clear_bit(num);
-    _mutator_free_bitmap.clear_bit(num);
-    // Touched the bounds? Need to update:
-    if (touches_bounds(num)) {
-      adjust_bounds();
+      assert(peek_is_mutator_free(idx), "Must be mutator free: " SIZE_FORMAT, idx);
+      clear_mutator_free(idx);
+      assert(!is_collector_free(idx) && !is_old_collector_free(idx), "Region cannot be in multiple free sets");
+      adjust_mutator_bounds_if_touched(idx);
+    } else if (r->free() < PLAB::min_size() * HeapWordSize) {
+      // Permanently retire this region if there's room for a fill object
+      size_t waste = r->free();
+      HeapWord* fill_addr = r->top();
+      size_t fill_size = waste / HeapWordSize;
+      if (fill_size >= ShenandoahHeap::min_fill_size()) {
+        ShenandoahHeap::fill_with_object(fill_addr, fill_size);
+        r->set_top(r->end());
+        // Since we have filled the waste with an empty object, account for increased usage
+        _heap->increase_used(waste);
+      } else {
+        waste = 0;              // if we don't make fill object, then the waste is not permanent
+      }
+      if (peek_is_old_collector_free(idx)) {
+        assert(_heap->mode()->is_generational(), "Old collector free regions only present in generational mode");
+        if (waste > 0) {
+          _heap->old_generation()->increase_used(waste);
+          _heap->card_scan()->register_object(fill_addr);
+        }
+        clear_old_collector_free(idx);
+        assert(!is_collector_free(idx) && !is_mutator_free(idx), "Region cannot be in multiple free sets");
+        adjust_old_collector_bounds_if_touched(idx);
+      } else if (peek_is_collector_free(idx)) {
+        if ((waste > 0) && _heap->mode()->is_generational()) {
+          _heap->young_generation()->increase_used(waste);
+        }
+        // This applies to both generational and non-generational mode
+        clear_collector_free(idx);
+        assert(!is_mutator_free(idx) && !is_old_collector_free(idx), "Region cannot be in multiple free sets");
+        adjust_collector_bounds_if_touched(idx);
+      }
```

Comment:
Do we ever get to generic `else` branch here? If not, we should assert that we are not expected to be there.